### ISSUE TITLE
Automatically Add Pod and Service CIDR Ranges to no_proxy Setting

### ISSUE THREAD

**Issue by ricfeatherstone** [openshift/openshift-ansible#1981]
When using the Global Proxy Options settings having to manually add the SDN and Service IP CIDR ranges to the openshift_no_proxy variable leads to possible duplication in the hosts file and issues if one is changed and not the other.

The hostnames are automatically added to this when the cluster is configured. It would be nice if these ranges, which are required, were automatically added for you

**Comment by sdodson** [comment]
@ricfeatherstone They should be added to the master config files, /etc/sysconfig/origin-master* or /etc/sysconfig/atomic-openshift-master* but they cannot be added to docker's config because docker doesn't respect CIDR ranges. Unfortunately that means that once you've created the registry you must add that ip address to `openshift_no_proxy`.

**Comment by detiber** [comment]
@sdodson should we "reserve" an IP for the registry? say the second address in CIDR block?

**Comment by ricfeatherstone** [comment]
@sdodson, @detiber 

OK, so there's a couple of points there. 

For clarity I'm doing an enterprise deployment so looking at the docs [here](https://docs.openshift.com/enterprise/3.2/install_config/http_proxies.html), where it states you need to add the SDN and Service IP CIDR ranges.

Is this is a requirement for master function, and this is refering specifically to the master config file(s)? Given those values are already specified in osm_cluster_network_cidr and openshift_master_portal_net or use defaults both of these values are already known. This means also adding them to openshift_no_proxy is redundant.

The second point is the docker configuration, and I understand that you can't put a CIDR range in the docker no_proxy but having to install the registry and then update the docker config for the registry's service ip feels wrong.

Is it possible to use the internal name - docker-registry.default.svc.cluster.local when pushing images to the registry?

Prior to 3.1 I had separate playbooks for managing all of this (to be fair adding the registry ip as well as name to the docker proxy) which I was hoping to junk as they had a bit of smell to them. Given that the installer creates the Docker registry for me, would it not be right for that to handle the proxy exclusion (whether name or ip based)?

**Comment by sdodson** [comment]
@ricfeatherstone 

That documentation is mostly oriented around configuring things manually post installation. If you're using version 3.0.80 or newer of the installer we'll do this for you automatically either by asking you for the proxy information in the quick installer or by setting `openshift_http_proxy`, `openshift_https_proxy` and `openshift_no_proxy` variables in your inventory file, the advanced install bits are documented [here.](https://docs.openshift.com/enterprise/3.2/install_config/install/advanced_install.html#advanced-install-configuring-global-proxy) This version of the installer will add those two CIDR ranges to the master's configuration files automatically, that's done [here.](https://github.com/openshift/openshift-ansible/blob/master/roles/openshift_master/templates/atomic-openshift-master.j2#L21) Adding the SDN CIDR range is a requirement for the horizontal pod autoscaling, `oc exec`, and `oc proxy` because the master makes connections directly to pod ips. I'm not entirely certain of the need for the kube svc CIDR, but if it becomes a need an external proxy would definitely not be able to reach that network.

Our intention was to switch to using the hostname for registry pushes however after I finished the work to configure all the pieces we discovered that there were other missing pieces preventing us from pushing to the registry via host name. https://github.com/openshift/origin/issues/8581 tracks those problems. I had hoped that would be a quick fix however they're saying it won't happen until at least 3.3. So I think it's time to consider alternative fixes. I was going to pursue pre-creating the docker-registry service as soon as the master became available. We could do it after the steps that automatically create the registry but those steps come after docker has been configured so we'd have to do another round through all hosts reconfiguring and restarting docker.
